crates/red_knot_python_semantic/src/types/type_ordering.rs

```diff
@@ -264,11 +269,44 @@ pub(super) fn union_elements_ordering<'db>(left: &Type<'db>, right: &Type<'db>)
         (Type::Dynamic(_), _) => Ordering::Less,
         (_, Type::Dynamic(_)) => Ordering::Greater,
 
-        (Type::Union(left), Type::Union(right)) => left.cmp(right),
+        (Type::Union(left), Type::Union(right)) => {
+            if left.elements(db).len() != right.elements(db).len() {
+                return left.elements(db).len().cmp(&right.elements(db).len());
+            }
+            // Lexicographically compare the elements of the two unions.
+            for (left, right) in left.elements(db).iter().zip(right.elements(db)) {
+                let ordering = union_elements_ordering(left, right, db);
+                if ordering != Ordering::Equal {
+                    return ordering;
+                }
+            }
+            Ordering::Equal
+        }
         (Type::Union(_), _) => Ordering::Less,
         (_, Type::Union(_)) => Ordering::Greater,
```

Comment:
Our representation of types is always in disjunctive normal form (union of intersections) and does not permit nested unions. Since this method is only used to sort elements of unions and intersections, it is impossible for us to ever encounter a union here. I think we should use `unreachable!` to assert this invariant, rather than adding dead code here.

```suggestion
        (Type::Union(_), _) | (_, Type::Union(_)) => {
            unreachable!("our type representation does not permit nested unions")
        }
```